Don't Use PW in ENV vars and default SSL on

fxdgear opened the pull request:
New Description to account for changes in direction during the life of this PR and feedback received. 

1. Don't use ENV vars to specify passwords for running containers 
2. Setup script and yml file will 
  * generate an `ELASTIC_PASSWORD` if one is not provided by the user at script execution time.
  * generate certs for security
  * generate keystores to store passwords securely 
3. Use docker secrets (aka a volume mount in compose) to securely include the sensitive info (keystores, certs, etc) in containers at container run time. 
4. Each configuration file is now included in the `config` directory and no longer are configs managed via env vars. 
  * this makes it easier to see what the configs are
  * make changes to them for personal use or experimentation
  * configs pull sensitive information from keystores previously generated.
5. Security is enabled by default and communication between services is done over ssl/tls.
  * with the exception of kibana. Kibana runs on port 5601 without ssl enabled. This is to avoid the hoops of navigating to a website with an untrusted (self signed) cert. But communication between Kibana and Elasticsearch are secured and use SSL.

xeraa commented on config/apm-server/apm-server.yml:
```diff
@@ -0,0 +1,17 @@
+  hosts: ['elasticsearch:9200']
+  protocol: "https"
+  username: elastic
+  # Read PW from auditbeat.keystore
```
Nitpick: `apm.keystore`

xeraa commented on config/filebeat/filebeat.yml:
```diff
@@ -0,0 +1,26 @@
+  hosts: ['elasticsearch:9200']
+  protocol: "https"
+  username: elastic
+  # Read PW from auditbeat.keystore
```
Ok, this must be a copy paste issue :)

fxdgear replied:
hah yeah it is I'll fix.

xeraa commented on config/filebeat/filebeat.yml:
```diff
@@ -0,0 +1,26 @@
+    reload.enabled: false
+
+processors:
+- add_cloud_metadata:
```
Shouldn't this be `add_docker_metadata`? Here and in the 3 other Beats configs

fxdgear replied:
Oh? Quite possibly. I just copied out the config from each container as it ran in the `master` branch. 

I wasn’t checking for correctness. 

But I suppose this exercise is a good one given that we can now analyze more easily the config for each service.

xeraa replied:
Ah, I didn't check that. I've created #40 to keep track of this — no point in adding that to this PR as well.

fxdgear replied:
+1

fxdgear commented:
@elasticdog Make is for Linux/Mac people. Windows users can use the multiple -f in docker compose. 

I’ll upstate to dynamically create keystores and certs. 

Thanks for the feedback.

jarpy commented:
There's something nasty happening with file permissions. When I bring up the stack I see this:

```
heartbeat            | Exiting: error loading config file: config file ("heartbeat.yml") can only be writable by the owner but the permissions are "-rw-rw-r--" (to fix the permissions use: 'chmod go-w /usr/share/heartbeat/heartbeat.yml')
heartbeat exited with code 1
metricbeat           | Exiting: error loading config file: config file ("metricbeat.yml") can only be writable by the owner but the permissions are "-rw-rw-r--" (to fix the permissions use: 'chmod go-w /usr/share/metricbeat/metricbeat.yml')
metricbeat exited with code 1
packetbeat           | Exiting: error loading config file: config file ("packetbeat.yml") can only be writable by the owner but the permissions are "-rw-rw-r--" (to fix the permissions use: 'chmod go-w /usr/share/packetbeat/packetbeat.yml')
auditbeat            | Exiting: error loading config file: config file ("auditbeat.yml") must be owned by the beat user (uid=0) or root
packetbeat exited with code 1
filebeat             | Exiting: error loading config file: config file ("filebeat.yml") can only be writable by the owner but the permissions are "-rw-rw-r--" (to fix the permissions use: 'chmod go-w /usr/share/filebeat/filebeat.yml')
filebeat exited with code 1
auditbeat exited with code 1
apm_server           | Exiting: error loading config file: config file ("apm-server.yml") can only be writable by the owner but the permissions are "-rw-rw-r--" (to fix the permissions use: 'chmod go-w /usr/share/apm-server/apm-server.yml')
```
Same behaviour on Windows and Linux.

jarpy commented:
I'm in two minds about running Kibana on HTTPS. It means that the first thing a brand new user sees is an error page with the word "unsafe" in it. _We_ know that's just a self-signed cert, but it doesn't give that "it just works" feeling.

fxdgear commented:
@elastic/infra if anyone would like to review this PR I'm happy to answer any questions. i think it's ready to go now for real world use.

tvernum commented:
> Keystores are pre-configured with the "changeme" password

Assuming I understand this correctly, can we please not do this.
We went to a lot of trouble to get rid of default passwords from ES, please don't add them back in via docker.

Ping: @elastic/es-security

fxdgear commented:
@tvernum 

What are you trying to understand. I'm creating a keystore and then adding the password "changeme" to it. 

If you're wanting to get away from the password "changeme" I can default it to something else. 

the other option is I can use an automatically generated password and output it in the setup script after everything is done. Something like:

```
The elastic password is: blahblahblah
```

Will that work?

tvernum commented:
@fxdgear Sorry, I had forgotten that stack-docker has historically had a pre-configured password ( 😢  😭 🤢 😡 )
 
I would truly love if we never shipped _anything_ that had any preconfigured password in it. Even for things that are intended to be examples rather than production ready - users take what we provide as a gold standard and do likewise.

If it's possible for the setup process to ask for a password, or generate a password that would be a wonderful step forwards.

fxdgear commented:
@tvernum Does the latest commit work for you?
https://github.com/elastic/stack-docker/pull/39/commits/e1306b4997b04fba39dda28453a84222ff27c511

thanks for taking the time to review this!

Crazybus commented on README.md:
```diff
@@ -3,21 +3,39 @@ This example Docker Compose configuration demonstrates many components of the
 Point a browser at [`http://localhost:5601`](http://localhost:5601) to see the results.
+> *NOTE*: Elasticsearch is now setup with self-signed certs.
+
+Log in with `elastic` / `changeme`.
```
This should be updated to explain that the autogenerated password needs to be used if you don't set it yourself.

fxdgear replied:
+1 thanks will do.

Crazybus commented on Makefile:
```diff
@@ -28,3 +28,62 @@ $(TARGETS:%=%-checkout):
+		centos:7  "/usr/local/bin/setup-permissions.sh"
+
+setup_users:
+	docker-compose run --no-deps -e ELASTIC_PASSWORD=changeme elasticsearch /usr/local/bin/setup-users.sh
```
This hardcoded password doesn't work anymore here. This will need to use the autogenerted password that you have now added.

Crazybus commented on Makefile:
```diff
@@ -28,3 +28,62 @@ $(TARGETS:%=%-checkout):
+	docker-compose run --no-deps -e ELASTIC_PASSWORD=changeme heartbeat /usr/local/bin/setup-beat.sh heartbeat
+
+setup_metricbeat:
+	docker-compose run --no-deps -e ELASTIC_PASSWORD=changeme metricbeat /usr/local/bin/setup-beat.sh metricbeat
```
I didn't dive into it but when I try to run `make setup_metricbeat` with the `ELASTIC_PASSWORD` updated to my autogenerated one I'm getting Read-only file system errors.

```
chmod: changing permissions of '/usr/share/metricbeat/metricbeat.yml': Read-only file system
```

fxdgear commented:
@Crazybus 💥 thanks for the feedback. I forgot that I modified the Makefile. I'm gonna set that back to it's normal thing cause I put all the setup in a docker container. There's no need to do the make file anymore.

fxdgear commented:
@jarpy @tvernum @xeraa @Crazybus 

You all have commented on this PR at some point. I'd like to invite you to review it again. 

I feel like I've addressed everyones comments and concerns, and would like to get this "done" with so it's not something I have lingering anymore :) 

Thanks for all your assistance with this.

jarpy commented:
I think this breaks on Windows because it depends on the docker Unix socket.

```
C:\Users\jarpy\src\stack-docker [nick/dont_use_pw_in_env ≡]
λ  docker-compose -f setup.yml up
WARNING: The PWD variable is not set. Defaulting to a blank string.
WARNING: The Docker Engine you're using is running in swarm mode.

Compose does not use swarm mode to deploy services to multiple nodes in a swarm. All containers will be scheduled on the current node.

To deploy your application across the swarm, use `docker stack deploy`.

Creating stack-docker_setup_1 ... error

ERROR: for stack-docker_setup_1  Cannot create container for service setup: b'Mount denied:\nThe source path "\\\\var\\\\run\\\\docker.sock:/var/run/docker.sock"\nis not a valid Windows path'

ERROR: for setup  Cannot create container for service setup: b'Mount denied:\nThe source path "\\\\var\\\\run\\\\docker.sock:/var/run/docker.sock"\nis not a valid Windows path'
ERROR: Encountered errors while bringing up the project.
```

fxdgear commented:
@jarpy ahh thanks for this. you have to specify some env vars for a windows user and I forgot to put those in the read me. I'll add thanks. :)

Crazybus commented on Makefile:
```diff
@@ -28,3 +28,4 @@ $(TARGETS:%=%-checkout):
 
 $(TARGETS:%=%-clean):
 	rm -rf stack/$(@:%-clean=%)
```
Can we make sure `make clean` also removes the keystores? Took me a while to figure out why this was causing me problems when trying to spin up a fresh cluster after doing a `make clean` and a `delete all my docker data with fire`

joshbressers commented on scripts/setup.sh:
```diff
@@ -4,6 +4,9 @@ chown 1000 -R "$confdir"
 find "$confdir" -type f -name "*.keystore" -exec chmod go-wrx {} \;
 find "$confdir" -type f -name "*.yml" -exec chmod go-wrx {} \;
 
+PW=$(date +%s | sha256sum | base64 | head -c 16 ;)
```
We shouldn't do this to generate the password, it will be predictable. Better to grab some data from /dev/urandom or use openssl

jordansissel commented:
Possible error conditions I've found:

On second attempts to `-f setup.yml up`, some files may already exist: 

1) When elasticsearch.keystore exists, it is not regenerated. Still, a new ELASTIC_PASSWORD is generated and used in Kibana and Logstash, but not set in elasticsearch.keystore. Workaround: Remove `elasticsearch.keystore` and `-f setup up` again.

2) Kibana's SSL files, when they already exist, may not be repopulated. Workaround: Remove all new/modified files in `config/` and `git checkout .` to restore normal files, then `-f setup up` again.

fxdgear commented:
@jordansissel if you have `make` installed you can run a `make clean` which will solve this as well. BUT unfortuanly make isn't installed by default on windows :( 


Thank you so much for all your help on windows testing!

russcam commented on README.md:
```diff
@@ -3,21 +3,51 @@ This example Docker Compose configuration demonstrates many components of the
+  * `PWD=/path/to/checkout/for/stack-docker`
+    * for example I use the path: `/c/Users/nick/elastic/stack-docker`
+  * You can set these two ways:
+    1. in powershell use: `$Env:COMPOSE_CONVERT_WINDOWS_PATHS=1`
```
This will only set the variable for the current PowerShell session, which may be OK for this? 

The equivalent of option 2. would be

```powershell
[Environment]::SetEnvironmentVariable("COMPOSE_CONVERT_WINDOWS_PATHS", "1", "Machine")
```
As per number 2, you'd need to start a new session after this, for the change to be picked up (or _refresh_ the environment from the registry).

fxdgear replied:
@russcam thanks! I had no idea you could do that! ;)

jordansissel commented:
Tested latest commit, and it works for me (Tested on Windows 10 / powershell).

fxdgear commented:
@Crazybus can you please review again. I hope I've addressed all your changes. 
@jordansissel +1 thanks for the patch super helpful. 
@russcam Can you please review the README again. i've taken your suggestion and added it. 
@jarpy When you have time lets sit down and see where you are blocked.

russcam commented:
The PowerShell environment variable change looks good @fxdgear 👍
